File: ruby/lib/clone_page_handler.rb
require 'json'
require 'net/http'

class ClonePageHandler
  def initialize(username, page)
    @username = username
    @page = page
  end

  def call
    return false unless clonable?

    Kernel.system(clone_command)
  end

  private

  attr_reader :username, :page

  def info_uri
    URI("https://api.github.com/users/#{username}/repos?page=#{page}")
  end

  def response
    @response ||= Net::HTTP.get_response(info_uri)
  end

  def response_body
    @response_body ||= begin
      return unless response.code == '200'

      JSON.parse(response.body)
    end
  end

  def clone_command
    @clone_command ||= response_body&.map do |info|
      "git clone #{info['clone_url']} #{username}/#{info['name']}"
    end&.join(' & ')
  end

  def clonable?
    !username.nil? &&
      username.length > 0 &&
      !clone_command.nil? &&
      !clone_command.empty?
  end
end

File: python/github_repo_cloner/clone_page_handler.py
import subprocess
import requests

def clone(username, page):
    if username in (None, ''):
        return False

    repositories = get_repositories(username, page)
    if len(repositories) == 0:
        return False

    commands = map(lambda repo_info: create_command(repo_info, username), repositories)
    final_command = ' & '.join(commands)
    subprocess.run(final_command, shell=True, check=False)
    return True

def create_command(repo_info, username):
    return f'git clone {repo_info["clone_url"]} {username}/{repo_info["name"]}'

def get_repositories(username, page):
    response = requests.get(f'https://api.github.com/users/{username}/repos?page={page}')
    if response.status_code != 200:
        return []

    return response.json()
